crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,194 @@
+            _ => {}
+        }
+        visitor::walk_stmt(self, stmt);
+    }
+}
+
+fn name_ranges_from_expr<'a, U>(target: &'a Expr<U>, locator: &Locator) -> Vec<Range> {
+    find_names(target, locator)
+}
+
+fn name_ranges_from_with_items<'a, U>(items: &'a [Withitem<U>], locator: &Locator) -> Vec<Range> {
+    items
+        .iter()
+        .filter_map(|item| {
+            item.optional_vars
+                .as_ref()
+                .map(|expr| find_names(&**expr, locator))
+        })
+        .flatten()
+        .collect()
+}
+
+fn name_ranges_from_assign_targets<'a, U>(targets: &'a [Expr<U>], locator: &Locator) -> Vec<Range> {
+    targets
+        .iter()
+        .flat_map(|target| find_names(target, locator))
+        .collect()
+}
+
+/// PLW2901
+pub fn redefined_loop_name<'a, 'b>(checker: &'a mut Checker<'b>, node: &Node<'b>)
+where
+    'b: 'a,
+{
+    let (outer_name_ranges, inner_name_ranges) = match node {
+        Node::Stmt(stmt) => match &stmt.node {
+            // With.
+            StmtKind::With { items, body, .. } => {
+                let name_ranges = name_ranges_from_with_items(items, checker.locator);
+                let mut visitor = InnerForWithAssignNamesVisitor {
+                    locator: checker.locator,
+                    name_ranges: vec![],
+                };
+                for stmt in body {
+                    visitor.visit_stmt(stmt);
+                }
+                (name_ranges, visitor.name_ranges)
+            }
+            // For and async for.
+            StmtKind::For {
+                target,
+                body,
+                iter: _,
+                orelse: _,
+                ..
+            }
+            | StmtKind::AsyncFor {
+                target,
+                body,
+                iter: _,
+                orelse: _,
+                ..
+            } => {
+                let name_ranges = name_ranges_from_expr(target, checker.locator);
+                let mut visitor = InnerForWithAssignNamesVisitor {
+                    locator: checker.locator,
+                    name_ranges: vec![],
+                };
+                for stmt in body {
+                    visitor.visit_stmt(stmt);
+                }
+                (name_ranges, visitor.name_ranges)
+            }
+            _ => panic!(
+                "redefined_loop_name called on Statement that is not a With, For, or AsyncFor"
+            ),
+        },
+        Node::Expr(_) => panic!("redefined_loop_name called on Node that is not a Statement"),
+    };
+
+    let outer_names: Vec<&str> = outer_name_ranges
+        .iter()
+        .map(|range| checker.locator.slice(range))
+        // Ignore dummy variables.
+        .filter(|name| !checker.settings.dummy_variable_rgx.is_match(name))
+        .collect();
+    let inner_names: Vec<&str> = inner_name_ranges
+        .iter()
+        .map(|range| checker.locator.slice(range))
+        .collect();
+
+    for outer_name in &outer_names {
```

Comment:
(Just out of interest, how closely does this match Pylint's own implementation? Did you look at the Pylint source, or is it implemented from scratch?)

Reply:
It was implemented from scratch, but after I learned Pylint has it too, I did look at the Pylint source later to compare. Pylint has a slightly more efficient algorithm, using a single recursion over the AST, keeping track of defined names using a stack, and pushing and popping the names on entering and leaving each for loop. This could probably be done with a deeper integration into `ast.rs`.